### PR TITLE
Expected risk minimization and denotation loss

### PR DESCRIPTION
**Pull request by pdasigi**
Summary of changes:
1) Added an `ExpectedRiskMinimization` decoder trainer that can generally be used for `encoder_decoder` models, and used that for training the nlvr parser.
2) Changed the checklists to contain hard counts of agenda items. They used to be soft counts earlier.
3) Next action probabilities also depend on current checklist. The way the checklist is incorporated into next action probability calculation is not very clean at this moment, but I'll fix it in another PR.
4) Defined a "cost" based on coverage and denotation losses. This combination can be improved too.
5) I made some changes to the rules for getting the agenda so that there are no "impossible-to-fulfill" agendas.

These changes do not improve the denotation accuracies yet, but the coverage metric goes up (to about 75% from less than 50% earlier), and the logical forms generally look much better. There are more fixes I can think of, like penalizing model for producing action not on the agenda, but I'll make separate PRs for those.

### CODE REVIEW

**Comment by pdasigi** [comment]
@matt-gardner, this should now be ready for review. Sorry this is long, but I'll try to make the PRs after this smaller.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py [hunk 1]
```diff
@@ -804,13 +993,15 @@ def _compute_new_states(cls,  # type: ignore
...
         # checklist scores later.
-        probs = torch.exp(log_probs)
-        # batch_index -> [(group_index, action_index, action, checklist, score)]
+        considered_action_probs = nn_util.masked_softmax(action_logits, action_mask)
+        all_agenda_action_probs = state.get_agenda_action_probs()
+        # batch_index -> [(group_index, action_index, action, checklist, model_score)]
         next_states_info: Dict[int, List[Tuple[int, int, int, Variable, Variable]]] = defaultdict(list)
-        for group_index, (batch_index, instance_action_probs) in enumerate(zip(state.batch_indices, probs)):
+        for group_index, (batch_index, instance_score, instance_action_probs) in \
+            enumerate(zip(state.batch_indices, state.score, considered_action_probs)):
```
Comment (matt-gardner) [inline allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: This formatting is pretty bad - it makes it look like the `enumerate` is the first line in the `for` block.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py [hunk 2]
```diff
@@ -820,25 +1011,31 @@ def _compute_new_states(cls,  # type: ignore
                     continue
                 # If action is not in instance_agenda, mask_variable, and checklist_addition will be
                 # all 0s.
-                checklist_mask = instance_agenda == action  # (agenda_size,)
-                checklist_addition = checklist_mask.float() * action_prob  # (agenda_size,)
+                checklist_mask = (instance_agenda == action).float()  # (agenda_size,)
+                agenda_action_prob = torch.sum(checklist_mask * agenda_action_probs)  # (1,)
+                # TODO (pradeep): This is not a great way to bias the model towards choosing actions
```
Comment (matt-gardner) [inline allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: One other option: see equation 11 here: https://arxiv.org/pdf/1704.04368.pdf.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py [hunk 3]
```diff
@@ -820,25 +1011,31 @@ def _compute_new_states(cls,  # type: ignore
...
+                # that fill the checklist. May be use the current checklist in the action logit
+                # computation itself.
+                action_prob = checklist_weight * agenda_action_prob + \
```
Comment (matt-gardner) [inline allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: Are you actually ending up with a valid probability distribution when you do this?  Also, my recommendation would be to put this logic in `_take_step`, and still have `log_probs` be the input to this function, so there's a more obvious distinction between where model computation happens and where state booking happens.

File: allennlp/nn/decoding/expected_risk_minimization.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    other one being sampling; see "Classical Structured Prediction Losses for Sequence to Sequence
+    Learning" by Edunov et al., 2017 for more details).
+    Note that we do not have a notion of targets here, so we're breaking the API of DecoderTrainer
+    a bit.
+
+    Parameters
+    ----------
+    beam_size : ``int``
+    noramlize_by_length : ``bool``
+        Should the log probabilities by normalized by length before renormalizing them? Edunov et
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/nn/decoding/expected_risk_minimization.py]: Done.

File: allennlp/nn/decoding/expected_risk_minimization.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                if not_finished is not None:
+                    next_states.append(not_finished)
+
+            states = self._prune_beam(next_states)
+            num_steps += 1
+
+        loss = nn_util.new_variable_with_data(initial_state.score[0], torch.Tensor([0.0]))
+        finished_model_scores = self._get_model_scores_by_batch(finished_states)
+        finished_costs = self._get_costs_by_batch(finished_states)
+        for index in finished_model_scores:
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/nn/decoding/expected_risk_minimization.py]: Done.

File: allennlp/nn/decoding/expected_risk_minimization.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            # Finished model scores are log-probabilities of the predicted sequences. We convert
+            # them into sequence probabilities and re-normalize them to compute expected cost under
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/nn/decoding/expected_risk_minimization.py]: I meant convert log-probabilities into probabilities. I changed the comment to make this clearer, and added another comment above the unmasked softmax to say what's happening there.

File: allennlp/nn/decoding/decoder_state.py [hunk 7]
```diff
@@ -53,6 +54,13 @@ def __init__(self,
         self.action_history = action_history
         self.score = score
 
+    def get_cost(self) -> Variable:
+        """
+        Returns cost associated with a finished ``DecoderState`` if the training algorithm used
+        calls for it.
+        """
+        raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/nn/decoding/decoder_state.py]: Done.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py [hunk 8]
```diff
@@ -109,14 +143,17 @@ def forward(self,  # type: ignore
         # Get a mapping from production rules to global action ids.
         agenda_mask = agenda != -1
         agenda_mask = agenda_mask.long()
+        # TODO (pradeep): Use an unindexed field for labels?
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: That wouldn't necessarily make things faster, and the current method is clearer. So I think I'll just keep it this way for now and change it in the future if I find that its too slow on the full dataset.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py [hunk 9]
```diff
@@ -479,13 +551,123 @@ def __init__(self,
...
 
-    def get_valid_actions(self) -> List[List[int]]:
+    def get_valid_actions(self, shuffle: bool = True) -> List[List[int]]:
         """
         This method is identical to ``WikiTablesDecoderState.get_valid_actions``.
         Returns a list of valid actions for each element of the group.
         """
-        return [state.get_valid_actions() for state in self.grammar_state]
+        valid_actions = [state.get_valid_actions() for state in self.grammar_state]
+        if shuffle:
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: When we have a fixed order for iterating over valid actions, the model is more likely to see the earlier actions than the later ones because there's a beam, especially in the case of non-agenda actions. I suspected that this may be biasing the model towards them. But shuffling did not result in noticeable changes in the accuracy.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py [hunk 10]
```diff
@@ -479,13 +551,123 @@ def __init__(self,
...
+                # In-place shuffle
+                numpy.random.shuffle(actions)
+        return valid_actions
+
+    def denotation_is_correct(self) -> bool:
+        """
+        Returns whether action history in the state evaluates to the correct denotation. Only
+        defined when the state is finished.
+        """
+        assert self.is_finished(), "Cannot compute denotions for unfinished states!"
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: Done.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: If you have a model score, not just a coverage score, this isn't true - the actions will get re-sorted by model score, which is initialized basically randomly.  There's no way that you'd end up with the same ordering on the beam as you saw here.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py [hunk 11]
```diff
@@ -479,13 +551,123 @@ def __init__(self,
...
+    def get_agenda_action_probs(self) -> List[Variable]:
+        """
+        Returns probabilities of choosing in-agenda actions based on how many times they've been
+        chosen previously. That is, if the checklist value for the first action is higher than that
+        for the second action, the returned probability for the second action will be higher than
+        that for the first action.
+        """
+        agenda_probs: List[Variable] = []
+        for instance_agenda_mask, instance_checklist in zip(self.agenda_mask, self.checklist):
+            float_mask = instance_agenda_mask.float()
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: Turns out I did not have to transpose twice. I found out a bug related to this after I submitted this PR. Fixed it here and added shape information.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py (hunk 1 above)
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: Done.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py (hunk 2 above)
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: Great! Thanks for the pointer. This is exactly what I meant by "use the current checklist in the action logit computation" in the comment there.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py (hunk 3 above)
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: Yes, it is a valid probability distribution. I agree with you that model computation should happen in `_take_step`. There are some changes in this method in the #952. I'll make the change you recommended there since that would be easier.
